#!/usr/bin/env python3
import os, sys, runpy, traceback

def main():
    success = True
    for root, _, files in os.walk('tests'):
        for file in files:
            if file.startswith('test_') and file.endswith('.py'):
                path = os.path.join(root, file)
                try:
                    runpy.run_path(path)
                    print(f"{path}: PASSED")
                except AssertionError as e:
                    print(f"{path}: FAIL ({e})")
                    success = False
                except Exception as e:
                    print(f"{path}: ERROR ({e})")
                    traceback.print_exc()
                    success = False
    sys.exit(0 if success else 1)

if __name__ == '__main__':
    main()
